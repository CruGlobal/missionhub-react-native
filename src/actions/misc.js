import { Linking } from 'react-native';

import { trackActionWithoutData } from './analytics';
import { getContactSteps } from './steps';
import { reloadJourney } from './journey';
import { createContactAssignment } from './person';
import { contactAssignmentSelector } from '../selectors/people';
import { navigatePush } from './navigation';
import { PERSON_STAGE_SCREEN } from '../containers/PersonStageScreen';

export function openCommunicationLink(url, action) {
  //if someone has a better name for this feel free to suggest.
  return dispatch =>
    Linking.canOpenURL(url)
      .then(supported => {
        if (!supported) {
          WARN("Can't handle url: ", url);
          return;
        }

        Linking.openURL(url)
          .then(() => {
            dispatch(trackActionWithoutData(action));
          })
          .catch(err => {
            if (url.includes('telprompt')) {
              // telprompt was cancelled and Linking openURL method sees this as an error
              // it is not a true error so ignore it to prevent apps crashing
            } else {
              WARN('openURL error', err);
            }
          });
      })
      .catch(err => WARN('An unexpected error happened', err));
}

export function loadStepsAndJourney({ id: personId }, { id: organizationId }) {
  return dispatch => {
    dispatch(getContactSteps(personId, organizationId));
    dispatch(reloadJourney(personId, organizationId));
  };
}

export function assignContactAndPickStage(personId, orgId, myId) {
  return async dispatch => {
    const { person: resultPerson } = await dispatch(
      createContactAssignment(orgId, myId, personId),
    );

    const { id: contactAssignmentId } = contactAssignmentSelector(
      { auth: { person: { id: myId } } },
      { person: resultPerson, orgId },
    );

    dispatch(
      navigatePush(PERSON_STAGE_SCREEN, {
        contactId: resultPerson.id,
        orgId: orgId,
        contactAssignmentId,
        name: resultPerson.first_name,
        onComplete: () => {},
        section: 'people',
        subsection: 'person',
      }),
    );
  };
}
