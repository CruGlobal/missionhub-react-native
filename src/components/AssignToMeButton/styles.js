import { StyleSheet } from 'react-native';

import theme from '../../theme';

export default StyleSheet.create({
  assignButton: {
    paddingVertical: 8,
    backgroundColor: theme.red,
    borderRadius: 25,
    minWidth: 200,
    alignSelf: 'center',
    marginVertical: 8,
  },
  assignButtonText: {
    fontSize: 14,
  },
});
