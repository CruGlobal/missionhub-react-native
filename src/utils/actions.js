/* eslint max-params: 0 */

import {
  DEFAULT_PAGE_LIMIT,
  RESET_CHALLENGE_PAGINATION,
  RESET_CELEBRATION_PAGINATION,
  GLOBAL_COMMUNITY_ID,
} from '../constants';
import callApi from '../actions/api';
import { REQUESTS } from '../api/routes';
import { organizationSelector } from '../selectors/organizations';

// Challenge and Celebrate action helpers

export const GET_CELEBRATE_INCLUDE =
  'subject_person.organizational_permissions,subject_person.contact_assignments';

export const CHALLENGE = 'challenge';
export const CELEBRATE = 'celebrate';

const getOrg = (orgId, getState) =>
  organizationSelector(
    { organizations: getState().organizations },
    { orgId: orgId },
  );

const getFeedType = (type, orgId) =>
  type === CHALLENGE
    ? REQUESTS.GET_GROUP_CHALLENGE_FEED
    : orgId === GLOBAL_COMMUNITY_ID
    ? REQUESTS.GET_GLOBAL_CELEBRATE_FEED
    : REQUESTS.GET_GROUP_CELEBRATE_FEED;

const getPaginationType = type =>
  type === CHALLENGE ? 'challengePagination' : 'celebratePagination';

// Query for the celebrate and challenge feeds
function buildQuery(type, orgId, page, personId) {
  return {
    page: {
      limit: DEFAULT_PAGE_LIMIT,
      offset: DEFAULT_PAGE_LIMIT * page,
    },
    ...(type === CELEBRATE ? { orgId, include: GET_CELEBRATE_INCLUDE } : {}),
    ...(type === CELEBRATE && personId
      ? { filters: { subject_person_ids: personId } }
      : {}),
    ...(type === CHALLENGE
      ? {
          filters: {
            organization_ids: orgId === GLOBAL_COMMUNITY_ID ? 'null' : orgId,
          },
          sort: '-active,-created_at',
        }
      : {}),
  };
}

const resetPaginationAction = (type, orgId) => ({
  type:
    type === CHALLENGE
      ? RESET_CHALLENGE_PAGINATION
      : RESET_CELEBRATION_PAGINATION,
  orgId,
});

// Use this for the challenge and celebrate feeds
export function getFeed(type, orgId, personId = null) {
  return (dispatch, getState) => {
    const org = getOrg(orgId, getState);
    const pagingType = getPaginationType(type);
    const { page, hasNextPage } = org[pagingType]
      ? org[pagingType]
      : { page: 0, hasNextPage: true };

    if (!hasNextPage) {
      // Does not have more data
      return Promise.resolve();
    }
    const query = buildQuery(type, orgId, page, personId);

    return dispatch(callApi(getFeedType(type, orgId), query));
  };
}

// Use this for reloading the challenge and celebrate feeds
export function reloadFeed(type, orgId) {
  return (dispatch, getState) => {
    const org = getOrg(orgId, getState);
    const pagingType = getPaginationType(type);

    if (org) {
      if (org[pagingType]) {
        dispatch(resetPaginationAction(type, orgId));
      }
      //Sometimes reload gets called when no items had been loaded previously.  Call getFeed anyway for first set.
      return dispatch(getFeed(type, orgId));
    }
    return Promise.resolve();
  };
}
