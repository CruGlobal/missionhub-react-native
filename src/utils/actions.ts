/* eslint max-params: 0 */
import { ThunkDispatch } from 'redux-thunk';

import {
  DEFAULT_PAGE_LIMIT,
  RESET_CHALLENGE_PAGINATION,
  RESET_CELEBRATION_PAGINATION,
  GLOBAL_COMMUNITY_ID,
} from '../constants';
import callApi from '../actions/api';
import { REQUESTS } from '../api/routes';
import { organizationSelector } from '../selectors/organizations';
import { OrganizationsState } from '../reducers/organizations';

// Challenge and Celebrate action helpers

export const GET_CELEBRATE_INCLUDE =
  'subject_person.organizational_permissions,subject_person.contact_assignments';

export const CHALLENGE = 'challenge';
export const CELEBRATE = 'celebrate';

const getOrg = (
  orgId: string,
  getState: () => { organizations: OrganizationsState },
) =>
  organizationSelector(
    { organizations: getState().organizations },
    { orgId: orgId },
  );

const getFeedType = (type: string, orgId: string) =>
  type === CHALLENGE
    ? REQUESTS.GET_GROUP_CHALLENGE_FEED
    : orgId === GLOBAL_COMMUNITY_ID
    ? REQUESTS.GET_GLOBAL_CELEBRATE_FEED
    : REQUESTS.GET_GROUP_CELEBRATE_FEED;

const getPaginationType = (type: string) =>
  type === CHALLENGE ? 'challengePagination' : 'celebratePagination';

// Query for the celebrate and challenge feeds
function buildQuery(
  type: string,
  orgId: string,
  page: number,
  personId: string | null,
) {
  return {
    page: {
      limit: DEFAULT_PAGE_LIMIT,
      offset: DEFAULT_PAGE_LIMIT * page,
    },
    ...(type === CELEBRATE ? { orgId, include: GET_CELEBRATE_INCLUDE } : {}),
    ...(type === CELEBRATE && personId
      ? { filters: { subject_person_ids: personId } }
      : {}),
    ...(type === CHALLENGE
      ? {
          filters: {
            organization_ids: orgId === GLOBAL_COMMUNITY_ID ? 'null' : orgId,
          },
          sort: '-active,-created_at',
        }
      : {}),
  };
}

const resetPaginationAction = (type: string, orgId: string) => ({
  type:
    type === CHALLENGE
      ? RESET_CHALLENGE_PAGINATION
      : RESET_CELEBRATION_PAGINATION,
  orgId,
});

// Use this for the challenge and celebrate feeds
export function getFeed(
  type: string,
  orgId: string,
  personId: string | null = null,
) {
  return (
    dispatch: ThunkDispatch<{}, {}, any>,
    getState: () => { organizations: OrganizationsState },
  ) => {
    const org = getOrg(orgId, getState);
    const pagingType = getPaginationType(type);
    const { page, hasNextPage } = org[pagingType]
      ? org[pagingType]
      : { page: 0, hasNextPage: true };

    if (!hasNextPage) {
      // Does not have more data
      return Promise.resolve();
    }
    const query = buildQuery(type, orgId, page, personId);

    return dispatch(callApi(getFeedType(type, orgId), query));
  };
}

// Use this for reloading the challenge and celebrate feeds
export function reloadFeed(type: string, orgId: string) {
  return (
    dispatch: ThunkDispatch<{}, {}, any>,
    getState: () => { organizations: OrganizationsState },
  ) => {
    const org = getOrg(orgId, getState);
    const pagingType = getPaginationType(type);

    if (org) {
      if (org[pagingType]) {
        dispatch(resetPaginationAction(type, orgId));
      }
      //Sometimes reload gets called when no items had been loaded previously.  Call getFeed anyway for first set.
      return dispatch(getFeed(type, orgId));
    }
    return Promise.resolve();
  };
}
