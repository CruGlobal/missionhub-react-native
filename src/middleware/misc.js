import { checkForUnreadComments } from '../actions/unreadComments';
import {
  NAVIGATE_FORWARD,
  STEPS_TAB,
  PEOPLE_TAB,
  GROUPS_TAB,
  TRACK_TAB,
} from '../constants';

export default function misc({ dispatch }) {
  return next => action => {
    const returnValue = next(action);

    switch (action.type) {
      case NAVIGATE_FORWARD:
      case TRACK_TAB:
        checkUnreadCommentsNotification(action, dispatch);
        break;
    }

    return returnValue;
  };
}

function checkUnreadCommentsNotification(action, dispatch) {
  if (
    action.routeName === STEPS_TAB ||
    action.routeName === PEOPLE_TAB ||
    action.routeName === GROUPS_TAB
  ) {
    dispatch(checkForUnreadComments());
  }
}
