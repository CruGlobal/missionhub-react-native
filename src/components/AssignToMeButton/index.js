import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { translate } from 'react-i18next';
import { connect } from 'react-redux';

import { Button } from '../common';
import { navigatePush } from '../../actions/navigation';
import { PERSON_STAGE_SCREEN } from '../../containers/PersonStageScreen';
import styles from './styles';
import { contactAssignmentSelector } from '../../selectors/people';
import { createContactAssignment } from '../../actions/person';

@translate()
class AssignToMeButton extends Component {
  assignToMe = async () => {
    const { dispatch, personId, orgId, myId } = this.props;

    const { person: resultPerson } = await dispatch(
      createContactAssignment(orgId, myId, personId),
    );

    const { id: contactAssignmentId } = contactAssignmentSelector(
      { auth: { person: { id: myId } } },
      { person: resultPerson, orgId },
    );

    dispatch(
      navigatePush(PERSON_STAGE_SCREEN, {
        contactId: resultPerson.id,
        orgId: orgId,
        contactAssignmentId,
        name: resultPerson.first_name,
        onComplete: () => {},
        section: 'people',
        subsection: 'person',
      }),
    );
  };

  render() {
    const { onPress, t } = this.props;

    return (
      <Button
        type="transparent"
        onPress={this.assignToMe}
        text={t('assignToMe').toUpperCase()}
        style={styles.assignButton}
        buttonTextStyle={styles.assignButtonText}
      />
    );
  }
}

const mapStateToProps = ({ auth }) => ({
  myId: auth.person.id,
});
export default connect(mapStateToProps)(AssignToMeButton);

AssignToMeButton.propTypes = {
  orgId: PropTypes.string.isRequired,
  personId: PropTypes.string.isRequired,
};
