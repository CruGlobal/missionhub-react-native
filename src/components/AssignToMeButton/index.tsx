import React from 'react';
import { useTranslation } from 'react-i18next';
import { connect } from 'react-redux-legacy';
import { ThunkDispatch } from 'redux-thunk';

import { Button } from '../common';
import { assignContactAndPickStage } from '../../actions/misc';

import styles from './styles';

interface AssignToMeButtonProps {
  organization: object;
  person: object;
  onComplete?: () => void;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  dispatch: ThunkDispatch<any, null, never>;
}

const AssignToMeButton = ({
  dispatch,
  person,
  organization,
  onComplete,
}: AssignToMeButtonProps) => {
  const { t } = useTranslation();

  const assignToMe = async () => {
    await dispatch(assignContactAndPickStage(person, organization));
    onComplete && onComplete();
  };
  return (
    <Button
      testID="AssignToMeButton"
      type="transparent"
      onPress={assignToMe}
      text={t('assignToMe').toUpperCase()}
      style={styles.assignButton}
      buttonTextStyle={styles.assignButtonText}
    />
  );
};

export default connect()(AssignToMeButton);
