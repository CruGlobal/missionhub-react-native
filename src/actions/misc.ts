/* eslint max-params: 0, max-lines-per-function: 0 */

import { Linking } from 'react-native';

import { contactAssignmentSelector } from '../selectors/people';
import {
  SELECT_MY_STAGE_FLOW,
  SELECT_PERSON_STAGE_FLOW,
  ADD_MY_STEP_FLOW,
  ADD_PERSON_STEP_FLOW,
} from '../routes/constants';
import { WARN } from '../utils/logging';
import { buildTrackingObj } from '../utils/common';

import { trackActionWithoutData } from './analytics';
import { getContactSteps } from './steps';
import { reloadJourney } from './journey';
import { createContactAssignment, getPersonScreenRoute } from './person';
import { navigatePush, navigateReplace } from './navigation';

export function openCommunicationLink(url, action) {
  //if someone has a better name for this feel free to suggest.
  return dispatch =>
    Linking.canOpenURL(url)
      .then(supported => {
        if (!supported) {
          WARN("Can't handle url: ", url);
          return;
        }

        Linking.openURL(url)
          .then(() => {
            dispatch(trackActionWithoutData(action));
          })
          .catch(err => {
            if ((url || '').includes('telprompt')) {
              // telprompt was cancelled and Linking openURL method sees this as an error
              // it is not a true error so ignore it to prevent apps crashing
            } else {
              WARN('openURL error', err);
            }
          });
      })
      .catch(err => WARN('An unexpected error happened', err));
}

export function loadStepsAndJourney(personId, organizationId) {
  return dispatch => {
    dispatch(getContactSteps(personId, organizationId));
    dispatch(reloadJourney(personId, organizationId));
  };
}

export function assignContactAndPickStage(person, organization) {
  return async (dispatch, getState) => {
    const auth = getState().auth;
    const authPerson = auth.person;
    const myId = auth.person.id;
    const orgId = organization.id;
    const personId = person.id;

    const { person: resultPerson } = await dispatch(
      createContactAssignment(orgId, myId, personId),
    );

    const contactAssignment = contactAssignmentSelector(
      { auth },
      { person: resultPerson, orgId },
    );

    dispatch(
      navigateReplace(
        getPersonScreenRoute(
          authPerson,
          resultPerson,
          organization,
          contactAssignment,
        ),
        {
          person: resultPerson,
          organization,
        },
      ),
    );

    dispatch(
      navigatePush(SELECT_PERSON_STAGE_FLOW, {
        personId: resultPerson.id,
        orgId,
        section: 'people',
        subsection: 'person',
      }),
    );
  };
}

export function navigateToStageScreen(
  personIsCurrentUser,
  person,
  contactAssignment,
  organization = {},
  firstItemIndex, //todo find a way to not pass this
) {
  return dispatch => {
    if (personIsCurrentUser) {
      dispatch(
        navigatePush(SELECT_MY_STAGE_FLOW, {
          selectedStageId: firstItemIndex,
          personId: person.id,
          section: 'people',
          subsection: 'self',
        }),
      );
    } else {
      dispatch(
        navigatePush(SELECT_PERSON_STAGE_FLOW, {
          selectedStageId: firstItemIndex,
          personId: person.id,
          orgId: organization.id,
          section: 'people',
          subsection: 'person',
        }),
      );
    }
  };
}

export function navigateToAddStepFlow(
  personIsCurrentUser,
  person,
  organization,
) {
  return dispatch => {
    const trackingParams = {
      trackingObj: buildTrackingObj(
        'people : person : steps : add',
        'people',
        'person',
        'steps',
      ),
    };

    if (personIsCurrentUser) {
      dispatch(
        navigatePush(ADD_MY_STEP_FLOW, {
          ...trackingParams,
          organization,
        }),
      );
    } else {
      dispatch(
        navigatePush(ADD_PERSON_STEP_FLOW, {
          ...trackingParams,
          contactName: person.first_name,
          personId: person.id,
          organization,
          createStepTracking: buildTrackingObj(
            'people : person : steps : create',
            'people',
            'person',
            'steps',
          ),
        }),
      );
    }
  };
}
