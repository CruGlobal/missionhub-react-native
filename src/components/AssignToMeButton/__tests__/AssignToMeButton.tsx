import React from 'react';
import { fireEvent } from 'react-native-testing-library';

import { renderWithContext } from '../../../../testUtils';
import { assignContactAndPickStage } from '../../../actions/misc';

import AssignToMeButton from '..';

jest.mock('../../../actions/misc');
jest.mock('../../../actions/person');
jest.mock('../../../selectors/people');
jest.mock('../../../actions/navigation');

const person = { id: '100', first_name: 'Roge' };
const organization = { id: '800' };
const props = {
  person: person,
  organization,
};

const assignResponse = { type: 'success' };

it('renders correctly', () => {
  renderWithContext(<AssignToMeButton {...props} />).snapshot();
});

describe('assignToMe', () => {
  (assignContactAndPickStage as jest.Mock).mockReturnValue(assignResponse);

  it('calls assignContactAndPickStage on press', async () => {
    const { getByTestId } = renderWithContext(<AssignToMeButton {...props} />);

    await fireEvent.press(getByTestId('AssignToMeButton'));

    expect(assignContactAndPickStage).toHaveBeenCalledWith(person);
  });

  it('calls assignContactAndPickStage and onComplete on press', async () => {
    const onComplete = jest.fn();
    const { getByTestId } = renderWithContext(
      <AssignToMeButton {...props} onComplete={onComplete} />,
    );

    await fireEvent.press(getByTestId('AssignToMeButton'));

    expect(assignContactAndPickStage).toHaveBeenCalledWith(person);
    expect(onComplete).toHaveBeenCalled();
  });
});
