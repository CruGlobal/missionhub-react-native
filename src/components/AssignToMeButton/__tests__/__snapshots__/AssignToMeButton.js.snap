// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Button
  buttonTextStyle={
    Object {
      "fontSize": 14,
    }
  }
  onPress={[Function]}
  style={
    Object {
      "alignSelf": "center",
      "backgroundColor": "#FF5532",
      "borderRadius": 25,
      "marginVertical": 8,
      "minWidth": 200,
      "paddingVertical": 8,
    }
  }
  text="ASSIGN TO ME"
  type="transparent"
/>
`;
